File: src/test/java/com/churchapp/entity/MemberTest.java
package com.churchapp.entity;


import com.churchapp.entity.enums.RoleType;
import org.junit.jupiter.api.Test;

import java.time.LocalDate;

import static org.junit.jupiter.api.Assertions.*;

class MemberTest {

    @Test
    void testMemberBuilder() {
        // Given & When
        Member member = Member.builder()
            .id(1)
            .firstName("John")
            .lastName("Doe")
            .email("john.doe@example.com")
            .phone("555-1234")
            .dateOfBirth(LocalDate.of(1990, 1, 1))
            .membershipDate(LocalDate.now())
            .active(true)
            .build();

        // Then
        assertEquals(1, member.getId());
        assertEquals("John", member.getFirstName());
        assertEquals("Doe", member.getLastName());
        assertEquals("john.doe@example.com", member.getEmail());
        assertEquals("555-1234", member.getPhone());
        assertEquals(LocalDate.of(1990, 1, 1), member.getDateOfBirth());
        assertTrue(member.getActive());
    }

    @Test
    void testMemberEqualsAndHashCode() {
        // Given
        Member member1 = Member.builder()
            .id(1)
            .firstName("John")
            .lastName("Doe")
            .email("john.doe@example.com")
            .build();

        Member member2 = Member.builder()
            .id(1)
            .firstName("John")
            .lastName("Doe")
            .email("john.doe@example.com")
            .build();

        Member member3 = Member.builder()
            .id(2)
            .firstName("Jane")
            .lastName("Doe")
            .email("jane.doe@example.com")
            .build();

        // Then
        assertEquals(member1, member2);
        assertNotEquals(member1, member3);
        assertEquals(member1.hashCode(), member2.hashCode());
    }

    @Test
    void testMemberToString() {
        // Given
        Member member = Member.builder()
            .id(1)
            .firstName("John")
            .lastName("Doe")
            .email("john.doe@example.com")
            .build();

        // When
        String result = member.toString();

        // Then
        assertNotNull(result);
        assertTrue(result.contains("John"));
        assertTrue(result.contains("Doe"));
    }

    @Test
    void testMemberSettersAndGetters() {
        // Given
        Member member = new Member();
        LocalDate birthDate = LocalDate.of(1990, 1, 1);
        LocalDate membershipDate = LocalDate.now();

        // When
        member.setId(1);
        member.setFirstName("John");
        member.setLastName("Doe");
        member.setEmail("john.doe@example.com");
        member.setPhone("555-1234");
        member.setDateOfBirth(birthDate);
        member.setMembershipDate(membershipDate);
        member.setActive(true);

        // Then
        assertEquals(1, member.getId());
        assertEquals("John", member.getFirstName());
        assertEquals("Doe", member.getLastName());
        assertEquals("john.doe@example.com", member.getEmail());
        assertEquals("555-1234", member.getPhone());
        assertEquals(birthDate, member.getDateOfBirth());
        assertEquals(membershipDate, member.getMembershipDate());
        assertTrue(member.getActive());
    }
}

File: src/test/kotlin/com/churchapp/entity/MemberTest.kt
package com.churchapp.entity

import org.junit.jupiter.api.Test
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Assertions.*
import java.time.LocalDate
import java.util.*

@DisplayName("Member Entity Tests")
class MemberTest {

    private lateinit var member: Member
    private lateinit var family: Family

    @BeforeEach
    fun setup() {
        family = Family(
            id = UUID.randomUUID(),
            familyName = "Smith Family"
        )

        member = Member.builder()
            .firstName("John")
            .lastName("Doe")
            .email("john.doe@example.com")
            .phoneNumber("123-456-7890")
            .address("123 Main St")
            .dateOfBirth(LocalDate.of(1990, 1, 1))
            .membershipDate(LocalDate.now())
            .build()
    }

    @Test
    fun `should create member with builder pattern`() {
        assertNotNull(member)
        assertEquals("John", member.firstName)
        assertEquals("Doe", member.lastName)
        assertEquals("john.doe@example.com", member.email)
        assertTrue(member.isActive)
    }

    @Test
    fun `should get full name correctly`() {
        assertEquals("John Doe", member.fullName)
    }

    @Test
    fun `should handle phone number option`() {
        assertTrue(member.getPhoneNumberOption().isDefined())
        assertEquals("123-456-7890", member.getPhoneNumberOption().getOrNull())
    }

    @Test
    fun `should handle missing phone number`() {
        val memberWithoutPhone = Member.builder()
            .firstName("Jane")
            .lastName("Doe")
            .email("jane@example.com")
            .build()
        
        assertTrue(memberWithoutPhone.getPhoneNumberOption().isEmpty())
    }

    @Test
    fun `should handle address option`() {
        assertTrue(member.getAddressOption().isDefined())
        assertEquals("123 Main St", member.getAddressOption().getOrNull())
    }

    @Test
    fun `should handle date of birth option`() {
        assertTrue(member.getDateOfBirthOption().isDefined())
        assertEquals(LocalDate.of(1990, 1, 1), member.getDateOfBirthOption().getOrNull())
    }

    @Test
    fun `should handle family relationship`() {
        val memberWithFamily = Member.builder()
            .firstName("Child")
            .lastName("Smith")
            .email("child@example.com")
            .family(family)
            .build()
        
        assertTrue(memberWithFamily.getFamilyOption().isDefined())
        assertEquals(family, memberWithFamily.getFamilyOption().getOrNull())
    }

    @Test
    fun `should handle missing family relationship`() {
        assertTrue(member.getFamilyOption().isEmpty())
    }

    @Test
    fun `should handle baptism date option`() {
        val baptizedMember = Member.builder()
            .firstName("Baptized")
            .lastName("Member")
            .email("baptized@example.com")
            .baptismDate(LocalDate.of(2020, 1, 1))
            .build()
        
        assertTrue(baptizedMember.getBaptismDateOption().isDefined())
        assertEquals(LocalDate.of(2020, 1, 1), baptizedMember.getBaptismDateOption().getOrNull())
    }

    @Test
    fun `should have default active status`() {
        assertTrue(member.isActive)
    }

    @Test
    fun `should handle inactive member`() {
        val inactiveMember = Member.builder()
            .firstName("Inactive")
            .lastName("Member")
            .email("inactive@example.com")
            .isActive(false)
            .build()
        
        assertFalse(inactiveMember.isActive)
    }

    @Test
    fun `should build member with all properties`() {
        val builtMember = Member.builder()
            .firstName("Jane")
            .lastName("Smith")
            .email("jane.smith@example.com")
            .phoneNumber("098-765-4321")
            .address("456 Oak St")
            .dateOfBirth(LocalDate.of(1992, 6, 15))
            .membershipDate(LocalDate.now())
            .baptismDate(LocalDate.of(2010, 1, 1))
            .family(family)
            .build()

        assertEquals("Jane", builtMember.firstName)
        assertEquals("Smith", builtMember.lastName)
        assertEquals("jane.smith@example.com", builtMember.email)
        assertEquals("098-765-4321", builtMember.getPhoneNumberOption().getOrNull())
        assertEquals("456 Oak St", builtMember.getAddressOption().getOrNull())
        assertEquals(family, builtMember.getFamilyOption().getOrNull())
    }
}
